mne/preprocessing/faster.py

```diff
@@ -0,0 +1,463 @@
+        for m in use_metrics:
+            s = metrics[m](data[:, chs])
+            b = find_outliers(s, thres, max_iter)
+            logger.info('\tBad by %s: %s' % (m, b))
+            bads[m].append(b)
+
+    bads = {k: np.concatenate(v).tolist() for k, v in bads.items()}
+    if return_by_metric:
+        return bads
+    else:
+        return np.unique(np.concatenate(list(bads.values()))).tolist()
+
+
+def faster_bad_components(ica, epochs, thres=3, use_metrics=None,
+                          power_gradient_range=None, max_iter=2,
+                          return_by_metric=False):
+    """Implements the third step of the FASTER algorithm.
+
+    This function attempts to automatically mark bad ICA components by
+    performing outlier detection.
+
+    Parameters
+    ----------
+    ica : Instance of ICA
+        The ICA operator, already fitted to the supplied Epochs object.
+    epochs : Instance of Epochs
+        The untransformed epochs to analyze.
+    thres : float
+        The threshold value, in standard deviations, to apply. A component
+        crossing this threshold value is marked as bad. Defaults to 3.
+    use_metrics : list of str
+        List of metrics to use. Can be any combination of:
+            'eog_correlation', 'kurtosis', 'power_gradient', 'hurst',
+            'median_gradient'
+        Defaults to all of them.
+    power_gradient_range : pair of floats
+        The (lower, upper) frequency limits of the power spectrum to use when
+        calculating the 'power_gradient' metric. In the FASTER paper, they set
+        these to the passband of the lowpass filter. Defaults to (25, 40).
+    max_iter : int
+        The maximum number of iterations performed during outlier detection
+        (defaults to 2, original FASTER paper uses 1).
+    return_by_metric : bool
+        Whether to return the bad components as a flat list (False, default) or
+        as a dictionary with the names of the used metrics as keys and the
+        bad components found by this metric as values.
+
+    Returns
+    -------
+    bads : list of int
+        The indices of the bad components.
+
+    See also
+    --------
+    ICA.find_bads_ecg
+    ICA.find_bads_eog
+    """
+    if power_gradient_range is None:
+        power_gradient_range = (25., 40.)
+
+    epochs = epochs.pick_channels(ica.info['ch_names'])
+    source_data = ica.get_sources(epochs).get_data().transpose(1, 0, 2)
+    source_data = source_data.reshape(source_data.shape[0], -1)
+
+    # Compute the transform matrix used by the ICA operator is necessary
+    if use_metrics is None or 'kurtosis' in use_metrics:
+        transform_matrix = np.dot(ica.mixing_matrix_.T,
+                                  ica.pca_components_[:ica.n_components_])
+        if hasattr(ica, '_pre_whitener') and ica.noise_cov is not None:
+            transform_matrix = np.dot(transform_matrix, ica._pre_whitener)
+
+    metrics = {
```

Comment:
Marijn, you need to rebase, otherwise I cannot make a reviewable PR. I
already rebased locally.

2015-07-09 9:43 GMT+02:00 Marijn van Vliet <notifications@github.com>:

> In mne/preprocessing/faster.py
> <https://github.com/mne-tools/mne-python/pull/1777#discussion_r34230411>:
>
> > +    """
> > +    if power_gradient_range is None:
> > +        power_gradient_range = (25., 40.)
> > +
> > +    epochs = epochs.pick_channels(ica.info['ch_names'])
> > +    source_data = ica.get_sources(epochs).get_data().transpose(1, 0, 2)
> > +    source_data = source_data.reshape(source_data.shape[0], -1)
> > +
> > +    # Compute the transform matrix used by the ICA operator is necessary
> > +    if use_metrics is None or 'kurtosis' in use_metrics:
> > +        transform_matrix = np.dot(ica.mixing_matrix_.T,
> > +                                  ica.pca_components_[:ica.n_components_])
> > +        if hasattr(ica, '_pre_whitener') and ica.noise_cov is not None:
> > +            transform_matrix = np.dot(transform_matrix, ica._pre_whitener)
> > +
> > +    metrics = {
>
> how's it coming with that PR @dengemann <https://github.com/dengemann> ?
> It doesn't have to be pretty, I can clean it up on my end.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1777/files#r34230411>.
>